### PR TITLE
 [12.0][MIG] web_translate_dialog + improvements

### PR DESCRIPTION
**Pull request by tschanzt**

### CODE REVIEW

**Comment by pedrobaeza** [comment]
Latest versions of this module were broken due to multiple ir.translation records when html translate was used. Have you taken this into account? And I think there's a rework in master about this. Maybe you can check it to see if it's worth to backport it.

**Comment by tschanzt** [comment]
@pedrobaeza No not actively, as i wasn't aware of the issue. But do you have more details on the issue maybe i can check it out.

**Comment by pedrobaeza** [comment]
#835

**Comment by tschanzt** [comment]
Well that part is rewritten a bit. But the technical details in #844 which tries to fix the same issue mentioned in #835 is because context isn't passed in kwargs which it now is.

**Comment by pedrobaeza** [comment]
OK, please check Travis status and I'll give it a try to see.

File: web_translate_dialog/static/src/js/web_translate_dialog.js [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+        $(window).resize();
+        this.initialize_html_fields(lang);
+    },
+    do_load_fields_values: function() {
+        var self = this,
+            deferred = [];
+
+        this.$el.find('.oe_translation_field').val('').removeClass('touched');
+        _.each(self.languages, function(lg) {
+            var context = session.user_context;
```
Comment (yvaucher) [inline web_translate_dialog/static/src/js/web_translate_dialog.js]: As discussed together, we want to change this line to create a new context for each language.

**Review by yvaucher — COMMENTED** [review]
Code review

LGTM, waiting for the change on session before approving.

File: web_translate_dialog/__manifest__.py [hunk 2]
```diff
@@ -0,0 +1,26 @@
+# Copyright 2012 Guewen Baconnier (Camptocamp SA)
+# Copyright 2016 Antonio Espinosa <antonio.espinosa@tecnativa.com>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+{
+    "name": "Web Translate Dialog",
+    "summary": "Easy-to-use pop-up to translate fields in several languages",
+    "version": "12.0.1.0.0",
+    "category": "Web",
+    "website": "https://odoo-community.org/",
```
Comment (simahawk) [inline web_translate_dialog/__manifest__.py]: https://github.com/OCA/web

**Review by Tardo — CHANGES_REQUESTED** [review]
If uses suggestion link gets this error:
error_translation

**Comment by tschanzt** [comment]
@Tardo i updated the pr. Should work now. 
@pedrobaeza do you want to give it a try as well

**Review by yvaucher — APPROVED** [review]

**Review by pedrobaeza — CHANGES_REQUESTED** [review]
Hi, I have been trying the module in runbot, and there's a weird behavior. I have loaded 2 languages: the English standard one + Spanish. Having the interface in Spanish, look what happens:

Peek 28-06-2019 16-55

**Comment by grindtildeath** [comment]
Small fixes: https://github.com/tschanzt/web/pull/1

**Comment by grindtildeath** [comment]
@tschanzt Small fixes here : https://github.com/tschanzt/web/pull/2

**Review by grindtildeath — APPROVED** [review]

**Comment by grindtildeath** [comment]
@pedrobaeza your issue is fixed in 2de0212  :wink:

**Comment by grindtildeath** [comment]
@Tardo can you please update your review ?

File: web_translate_dialog/static/src/js/web_translate_dialog.js [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+var rpc = require('web.rpc');
+var dom = require('web.dom');
+
+var _t = core._t;
+var QWeb = core.qweb;
+var Mutex = concurrency.Mutex;
+
+var translateDialog = Dialog.extend({
+    template: "TranslateDialog",
+    init: function(parent, event_data, res_id, content) {
```
Comment (Tardo) [inline web_translate_dialog/static/src/js/web_translate_dialog.js]: 'content' is really used?

File: web_translate_dialog/static/src/js/web_translate_dialog.js [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+    },
+    on_button_translate: function() {
+        var self = this;
+        $.when(this.has_been_loaded).then(function() {
+            self.open_translate_dialog();
+        });
+    },
+});
+
+View.include({
```
Comment (Tardo) [inline web_translate_dialog/static/src/js/web_translate_dialog.js]: :fire:

File: web_translate_dialog/static/src/js/web_translate_dialog.js [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+});
```
Comment (Tardo) [inline web_translate_dialog/static/src/js/web_translate_dialog.js:254]: :fire:

File: web_translate_dialog/static/src/xml/base.xml [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+                </th>
+            </tr>
+            <t t-foreach="widget.translatable_fields" t-as="field_name">
+              <t t-set="field" t-value="widget.view.searchView.fields[field_name]" />
+              <tr t-att-data-field="field_name">
+                  <td class="oe_form_frame_cell" width="1%" nowrap="nowrap">
+                      <label class="oe_label"><t t-esc="field.string"/>:</label>
+                  </td>
+                  <td t-foreach="widget.languages" t-as="lg" class="oe_form_frame_cell">
+                      <input t-if="field.type == 'char' || field.type == 'url'"
```
Comment (Tardo) [inline web_translate_dialog/static/src/xml/base.xml]:
```suggestion
                      <input t-if="['char','url'].indexOf(field.type) !== -1"
```

File: web_translate_dialog/static/src/xml/base.xml [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+                             type="text" t-attf-name="#{lg.code}-#{field_name}"
+                             value="" data-value="" class="oe_translation_field"/>
+                      <textarea t-if="field.type == 'text'"
```
Comment (Tardo) [inline web_translate_dialog/static/src/xml/base.xml]:
```suggestion
                      <textarea t-elif="field.type === 'text'"
```

File: web_translate_dialog/static/src/xml/base.xml [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+                                t-attf-name="#{lg.code}-#{field_name}" data-value=""
+                                class="oe_translation_field" ></textarea>
+                      <div t-if="field.type == 'html'" class="oe_form_field_html">
```
Comment (Tardo) [inline web_translate_dialog/static/src/xml/base.xml]:
```suggestion
                      <div t-elif="field.type === 'html'" class="oe_form_field_html">
```

**Review by Tardo — CHANGES_REQUESTED** [review]

**Comment by TDu** [comment]
:+1:

File: web_translate_dialog/static/src/js/web_translate_dialog.js [hunk 9]
```diff
@@ -0,0 +1,224 @@
...
+var _t = core._t;
+var QWeb = core.qweb;
+var Mutex = concurrency.Mutex;
+
+var translateDialog = Dialog.extend({
+    template: "TranslateDialog",
+    init: function(parent, options) {
+        this._super(parent,
+                    {title: _t("Translations"),
+                     width: '90%',
```
Comment (Tardo) [inline web_translate_dialog/static/src/js/web_translate_dialog.js]: I can't see where 'width' and 'height' are used...

File: web_translate_dialog/static/src/js/web_translate_dialog.js [hunk 10]
```diff
@@ -0,0 +1,224 @@
...
+        this.languages = langs;
+        this.languages_loaded.resolve();
+    },
+    open: function() {
+        // the template needs the languages
+        return $.when(this.languages_loaded).then($.proxy(this._super, this));
+    },
+    start: function() {
+        var self = this;
+        this.$el.find('.oe_translation_field').change(function() {
```
Comment (Tardo) [inline web_translate_dialog/static/src/js/web_translate_dialog.js]: All widgets have "$" method that call to 'this.$el.find'. You can replace all ```this.$el.find(``` by ```this.$(```

**Review by Tardo — CHANGES_REQUESTED** [review]

File: web_translate_dialog/static/src/xml/base.xml [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+<!-- Copyright 2012 Guewen Baconnier (Camptocamp SA)
+     Copyright 2016 Antonio Espinosa <antonio.espinosa@tecnativa.com>
+     License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl). -->
+<template>
+
+    <t t-name="TranslateDialog">
+        <div class="modal-body">
+            <table t-if="widget.translatable_field"
+                   class="oe_frame oe_forms oe_translation_form"
+                   border="0" cellpadding="0" cellspacing="0" width="100%">
```
Comment (Tonow-c2c) [inline web_translate_dialog/static/src/xml/base.xml]: in the future it's will be great to create a less file and remove style on xml

File: web_translate_dialog/static/src/js/web_translate_dialog.js [hunk 12]
```diff
@@ -71,6 +71,22 @@ var translateDialog = Dialog.extend({
 
         this.do_load_fields_values();
     },
+    resize_textareas: function(){
+        var textareas = this.$('textarea.oe_translation_field');
+        var max_height = 100
+        // Resize textarea either to the max height of its content if it stays
+        // in the modal or to the max height available in the modal
+        if (textareas !== undefined && textareas[0] !== undefined) {
```
Comment (simahawk) [inline web_translate_dialog/static/src/js/web_translate_dialog.js]: just check `$textareas.length`

**Comment by simahawk** [comment]
just a small chance on OCA PR, but looks good

File: web_translate_dialog/models/res_lang.py [hunk 13]
```diff
@@ -0,0 +1,19 @@
...
+
+    tr_sequence = fields.Integer(
+        string='Translation sequence',
+        help='Defines the order of language to appear in translation dialog',
+        default=10,
+    )
+
+    @api.model
+    def search(self, args, offset=0, limit=None, order=None, count=False):
+        print(order)
```
Comment (nikul-serpentcs) [inline web_translate_dialog/models/res_lang.py]: @tschanzt Removed `print`

Comment (grindtildeath) [reply web_translate_dialog/models/res_lang.py]: Whole extension of search is unneeded, thanks for the fast feedback :wink:

**Comment by grindtildeath** [comment]
@pedrobaeza @Tardo @simahawk @Tonow-c2c @nikul-serpentcs 
Can you guys please update your reviews before we rebase and squash ?

**Comment by grindtildeath** [comment]
@pedrobaeza @Tardo @simahawk @Tonow-c2c @nikul-serpentcs
ping, let's get this merged please :pray:

**Comment by pedrobaeza** [comment]
Sorry, not able to review now

File: web_translate_dialog/static/src/js/web_translate_dialog.js [hunk 14]
```diff
@@ -0,0 +1,241 @@
...
+var FormView = require('web.FormView');
+var View = require('web.AbstractView');
+var session  = require('web.session');
+var rpc = require('web.rpc');
+
+var _t = core._t;
+var QWeb = core.qweb;
+var Mutex = concurrency.Mutex;
+
+var translateDialog = Dialog.extend({
```
Comment (simahawk) [inline web_translate_dialog/static/src/js/web_translate_dialog.js]: TranslateDialog

File: web_translate_dialog/static/src/js/web_translate_dialog.js [hunk 15]
```diff
@@ -0,0 +1,241 @@
...
+
+        this.initialize_html_fields(lang);
+    },
+    do_load_fields_values: function() {
+        var self = this,
+            deferred = [];
+
+        this.$('.oe_translation_field').val('').removeClass('touched');
+
+        var deff = $.Deferred();
```
Comment (simahawk) [inline web_translate_dialog/static/src/js/web_translate_dialog.js]: def

File: web_translate_dialog/static/src/xml/base.xml [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+            <tr>
+                <th t-foreach="widget.languages" align="left">
+                    <div class="separator horizontal"><t t-esc="name"/></div>
+                </th>
+            </tr>
+            <t t-value="widget.translatable_field" t-set="field_name" />
+              <t t-set="field" t-value="widget.view.searchView.fields[field_name]" />
```
Comment (simahawk) [inline web_translate_dialog/static/src/xml/base.xml]: bad indent

File: web_translate_dialog/models/base.py [hunk 17]
```diff
@@ -0,0 +1,34 @@
...
+            rec.get('id'): {'en_US': (0, rec.get(field_name))}
+            for rec in read_res
+        }
+        for rec_id, values in res.items():
+            tr_read_res = self.env['ir.translation'].search_read([
+                ('name', '=', '%s,%s' % (self._name, field_name)),
+                ('res_id', '=', rec_id)
+            ])
+            for tr_res in tr_read_res:
+                if tr_res.get('lang') == 'en_US':
```
Comment (simahawk) [inline web_translate_dialog/models/base.py]: this could be excluded in the query

**Review by simahawk — APPROVED** [review]
LG, just few non-blocking comments.

**Comment by simahawk** [comment]
@grindtildeath please squash some commits :)

**Comment by tschanzt** [comment]
@simahawk updated and i squashed some commits

File: web_translate_dialog/models/base.py [hunk 18]
```diff
@@ -0,0 +1,33 @@
+# Copyright 2019 Camptocamp SA
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl)
+from odoo import models, api
+
+
+class BaseModel(models.BaseModel):
+
+    _inherit = 'base'
```
Comment (daramousk) [inline web_translate_dialog/models/base.py:8]: No need for this:

https://github.com/OCA/OCB/blob/12.0/odoo/models.py#L434

File: web_translate_dialog/static/src/js/web_translate_dialog.js [hunk 19]
```diff
@@ -0,0 +1,241 @@
...
+        this.close();
+    },
+    on_button_close: function() {
+        this.close();
+    },
+
+});
+
+FormView.include({
+    render_sidebar: function($node) {
```
Comment (daramousk) [inline web_translate_dialog/static/src/js/web_translate_dialog.js]: This does not seem to exist on the `FormView`, therefore not called anywhere.

File: web_translate_dialog/readme/USAGE.rst [hunk 20]
```diff
@@ -0,0 +1,8 @@
+Go to an object that has translatable fields (*Products* for example) and
+select *More > Translate* (or click on *Edit* and then click on the flag at
```
Comment (daramousk) [inline web_translate_dialog/readme/USAGE.rst]: I do not see a `More` button here as well.

**Review by daramousk — CHANGES_REQUESTED** [review]

File: web_translate_dialog/static/src/js/web_translate_dialog.js (hunk 19 above)
Comment (simahawk) [reply web_translate_dialog/static/src/js/web_translate_dialog.js]: @tschanzt it's `renderSidebar` indeed. Do you have any piece of code defining `render_sidebar`?

File: web_translate_dialog/models/base.py (hunk 18 above)
Comment (simahawk) [reply web_translate_dialog/models/base.py:8]: that's added only when current name is not `base`. In any case: explicit it's better than implicit :wink:

**Review by simahawk — CHANGES_REQUESTED** [review]
form sidebar needs a fix

**Comment by tschanzt** [comment]
@simahawk @daramousk i updated the sidebar stuff and readded possibility to actually use that function

**Comment by simahawk** [comment]
@daramousk would you mind to pass by again?

File: web_translate_dialog/static/src/xml/base.xml [hunk 21]
```diff
@@ -0,0 +1,53 @@
...
+            </div>
+          </td>
+          </tr>
+      </t>
+      </table>
+    </div>
+  </t>
+
+  <t t-name="TranslateDialog.buttons">
+    <button class="btn btn-sm oe_button btn-primary oe_form_translate_dialog_save_button">
```
Comment (MeetKD) [inline web_translate_dialog/static/src/xml/base.xml:45]: Can we remove oe_button here and btn-primary class.

File: web_translate_dialog/static/src/xml/base.xml [hunk 22]
```diff
@@ -0,0 +1,53 @@
...
+      <span>Save</span>
+    </button>
+    <button class="btn btn-sm oe_button btn-default oe_form_translate_dialog_cancel_button">
+      <span>Cancel</span>
```
Comment (MeetKD) [inline web_translate_dialog/static/src/xml/base.xml:49]: Same her.e btn-secondary Class may be ?

File: web_translate_dialog/static/src/xml/base.xml [hunk 23]
```diff
@@ -0,0 +1,53 @@
...
+            <div t-elif="field.type == 'html'" class="oe_form_field_html">
+              <textarea class="oe_translation_field oe_form_field"
+                        t-attf-name="#{lg.code}-#{field_name}" data-value=""/>
+            </div>
+          </td>
+          </tr>
+      </t>
+      </table>
+    </div>
+  </t>
```
Comment (MeetKD) [inline web_translate_dialog/static/src/xml/base.xml:42]: I would also suggest to replace all the oe instance like oe_label, oe_form_field etc.

**Review by MeetKD — COMMENTED** [review]
Some non-blocking improvements

**Comment by yvaucher** [comment]
@simahawk @pedrobaeza I reseted your review, but seems we have enough approval.

By the way @tschanzt doesn't work anymore at C2C, so I propose to proceed with this PR and make improvement in further PRs.

**Pull request merged**